power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -0,0 +1,471 @@
+            } else {
+                throw SerializationError{"An element can only be a list or dictionary!\n"};
+            }
+        }
+        element_number_ = -1;
+    }
+
+    void parse_array_element(void* element_pointer, msgpack::object const& obj,
+                             std::span<MetaAttribute const* const> attributes) const {
+        auto const arr = obj.as<ArraySpan>();
+        if (arr.size() != attributes.size()) {
+            throw SerializationError{
+                "An element of a list should have same length as the list of predefined attributes!\n"};
+        }
+        for (attribute_number_ = 0; attribute_number_ != static_cast<Idx>(attributes.size()); ++attribute_number_) {
+            parse_attribute(element_pointer, arr[attribute_number_], *attributes[attribute_number_]);
+        }
+        attribute_number_ = -1;
+    }
+
+    void parse_map_element(void* element_pointer, msgpack::object const& obj, MetaComponent const& component) const {
+        for (msgpack::object_kv const& kv : obj.as<MapSpan>()) {
+            attribute_key_ = key_to_string(kv);
+            Idx const found_idx = component.find_attribute(attribute_key_);
+            if (found_idx < 0) {
+                attribute_key_ = "";
+                continue; // allow unknown key for additional user info
+            }
+            parse_attribute(element_pointer, kv.val, component.attributes[found_idx]);
+        }
+        attribute_key_ = "";
+    }
+
+    void parse_attribute(void* element_pointer, msgpack::object const& obj, MetaAttribute const& attribute) const {
+        // skip for none
+        if (obj.is_nil()) {
+            return;
+        }
+        // call relevant parser
+        switch (attribute.ctype) {
+        case CType::c_double:
+            return parse_attribute_per_type<double>(element_pointer, obj, attribute);
+        case CType::c_double3:
+            return parse_attribute_per_type<RealValue<false>>(element_pointer, obj, attribute);
+        case CType::c_int8:
+            return parse_attribute_per_type<int8_t>(element_pointer, obj, attribute);
+        case CType::c_int32:
+            return parse_attribute_per_type<int32_t>(element_pointer, obj, attribute);
+        default:
+            throw SerializationError{"Unknown data type for attriute!\n"};
+        }
+    }
+
+    template <class T>
+    void parse_attribute_per_type(void* element_pointer, msgpack::object const& obj,
+                                  MetaAttribute const& attribute) const {
+        obj >> *reinterpret_cast<T*>(reinterpret_cast<char*>(element_pointer) + attribute.offset);
+    }
+
+    void handle_error(std::exception& e) const {
+        std::stringstream ss;
+        ss << e.what();
+        if (!root_key_.empty()) {
+            ss << "Position of error: " << root_key_;
+            root_key_ = "";
+        }
+        if (is_batch_ && scenario_number_ >= 0) {
+            ss << "/" << scenario_number_;
```

Comment:
maybe explicitly mention that this is scenario `xyz`. Same below.

Alternatively instead of using `"/" << xyz`, you can use `"[" << xyz << "]"`. Users will understand that it represents a key/index (since it's already a JSON-like format anyway), and actually querying that key/index in e.g. python will yield the exact result the users need to find the issue

Reply:
I do not understand your point. @mgovers can you elaborate?

Reply:
some error messagelike

```txt
an error occurred while parsing data["foo"][1]["bar"]
```

or

```txt
an error occurred while parsing data: component 'foo', scenario 1, attribute 'bar'
```

may be more readable to the user than

```txt
an error occurred while parsing data/foo/1/bar
```

Reply:
in addition, if the user wants to debug, that person can open e.g. in python like

```py
import json

with open('foo.json', encoding='utf-8') as f:
    data = json.load(f)

print(data["foo"][1]["bar"])
```

and see what went wrong

Reply:
Ah, now I understand. I copy this design from `nlohmann_json`. They use `json_pointer` with `/` as separator:
https://json.nlohmann.me/features/json_pointer/#introduction

What do you think?

Reply:
OHHH i get it now, you're right, they're using HTML-like query syntax: https://www.npmjs.com/package/json-pointer

Reply:
So shall we also keep like this?

Reply:
yes